### ISSUE TITLE
Command-line init scripts not executed against `:buildSrc` project

### ISSUE THREAD

**Issue by wolfs** [gradle/gradle#1055]
Currently, when passing an init script via the `-I` command line parameter to a Gradle build does not apply it to the `buildSrc` project.

### Expected Behavior
```
./gradlew -I init.gradle build
```
applies the init script `init.gradle` to `buildSrc` and the main build.

### Current Behavior
`init.gradle` is only applied to the main build.

### Context
 This is a problem when one wants to configure something consistently for both `buildSrc` and the main build. This would mostly happen on CI, when you are not able to add init scripts to Gradle user home.
This was going to be attacked by composite builds when a build with `buildSrc` was going to be treated as a composite. But we probably should to fix that before the composite build solution is in place.

### Steps to Reproduce (for bugs)
See above.

### Your Environment
Gradle 3.2.1 (and earlier).

**Comment by lptr** [comment]
This issue is annoying, and confusing to users, so we should fix this in any case. However, fixing this for 4.0 has some more serious importance for task output caching.

We are going to roll out task output caching in 4.0, and the way to configure the cache is to use init scripts. Most people will probably want to use an init script in their `~/.gradle/init.d` directory, but it also makes sense—especially for people trying out the feature—to pass a cache-enabling init script on the command-line.

Not considering these init scripts when building `buildSrc` means that the `buildSrc` build will not be cached. This has a direct impact, as everybody needs to build `buildSrc` after every change, which is of course a time penalty we could avoid by sharing the built `buildSrc` output between machines. However, there is another, more serious implication, too.

There are some unrelated issues with the Groovy compiler etc. that cause the compiled `buildSrc` output to be non-reproducible across different machines. Because the `buildSrc` output appears on the buildscript classpath, these differences will cause every machine to generate different cache keys for every task, preventing any cache reuse between machines. This is a bad thing as it essentially cancels any benefits of the shared cache.

A usual way to work around such issues is to cache `buildSrc` itself, so each machine will use the same cached `buildSrc` output binaries. This works nicely when the cache is enabled via an init script in `~/.gradle/init.d`, but fails when the init script is passed on the command-line.

Basically, the effect is that someone who wants to try task output caching would pass a cache-enabling init script on the command line, and expect to see their build fully cached on some other machine. What they experience instead is no caching at all, yet they'd see no warning that they are doing something wrong. Even though we knew about this problem for some time ourselves, it didn't prevent us from falling into this trap a number of times. Users eager to try out task output caching would be much more likely to experience disappointing results about task output caching due to this bug.

**Comment by kennknowles** [comment]
As I understand it, init scripts are the recommended way to set up custom private repositories noninvasively. Does this mean that the `buildSrc` approach cannot be used in such a scenario? You also cannot set the dependencies for a script plugin via init script. So what is left? Every plugin has to be built standalone?

**Comment by lukecwik** [comment]
Note that by specifying the init script inside the Gradle home `init.d` folder, the script was being able to be applied to the `buildSrc` directory for me in Gradle 4.8.1. I found this workaround due to a users bug report that was closed as wontfix and not migrated to Gradle issues https://issues.gradle.org/browse/GRADLE-3197.

**Comment by ysmaoui** [comment]
any updates about this issue?

can we pass an init script to the buildSrc project somehow?
I am using a gradle wrapper so no GRADLE_HOME ..

**Comment by Fleshgrinder** [comment]
Not stale, this issue is still present.

**Comment by lacasseio** [comment]
Passed the last couple of hours hitting this issue. It is still confusing for everyone.

**Comment by DPUkyle** [comment]
I just lost a few hours to this too :)

**Comment by Mochis** [comment]
Same here...

**Comment by adammurdoch** [comment]
This is fixed now, via https://github.com/gradle/gradle/pull/22540
